### ISSUE TITLE
Add "CIRCULAR" mode padding to tensorflow.pad()

### ISSUE THREAD

**Issue by st0nedB** [tensorflow/tensorflow#41894]
<em>Please make sure that this is a feature request. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:feature_template</em>

**System information**
- TensorFlow version (you are using): 2.2.0-rc3
- Are you willing to contribute it (Yes/No): Yes

**Describe the feature and the current behavior/state.**
The requested feature would enable tensorflow.pad() to perform circular padding, which is of importance in several signal processing applications.
Following the example in the [documentation for tf.pad()](https://www.tensorflow.org/api_docs/python/tf/pad) for other modes ("CONSTANT", "SYMMETRIC", "REFLECT"), the circular padding should result in a tensor like this:
```python
t = tf.constant([[1, 2, 3], [4, 5, 6]])
paddings = tf.constant([[1, 1,], [2, 2]])

tf.pad(t, paddings, "CIRCULAR")  # [[5, 6, 4, 5, 6, 4, 5],
                                 #  [2, 3, 1, 2, 3, 1, 2],
                                 #  [5, 6, 4, 5, 6, 4, 5],
                                 #  [2, 3, 1, 2, 3, 1, 2]]
```
**Will this change the current api? How?**
I think it should not, as this will simply be another option for the mode argument. Of course, the backend must be modified to provide the functionality.

**Who will benefit with this feature?**
As stated before, circular padding often occurs in signal processing applications. Another example would be images that stem from e.g. a cylindric or circular domain. 
IMHO this feature would enable higher-level APIs, e.g. the convolutional layers in Keras to add an option for "circular" padding, which in turn should enable easier use for the specific tasks mentioned above. Making the feature available to the higher-level API will benefit the mentioned signal processing applications.

**Any Other info.**
The functionality is implemented in Pytorch, where the [padding mode for Convolutional layers](https://pytorch.org/docs/master/generated/torch.nn.Conv1d.html#torch.nn.Conv1d) can be set to `circular`.
> padding_mode (string, optional) – 'zeros', 'reflect', 'replicate' or 'circular'. Default: 'zeros'

**Comment by maxrohleder** [comment]
Would also very much appreciate that! As @st0nedB mentioned, this would make life so much easier for people working e.g. in medical data science.

**Comment by Hendekagon** [comment]
what's the latest on this ? Is there a workaround/alternative meanwhile ?

**Comment by jaygala223** [comment]
Hi, I am new here and am going through the thread on this page and the documentation for `pad`. I would like to work on this issue

**Comment by straw-hat-coder** [comment]
Hi @Hendekagon, I've tried to understand how this works and there are a couple of ways through which i think we can achieve "CIRCULAR" mode padding using existing TensorFlow operations. One approach could be to use the `tf.roll()` function to perform the circular shift of the input tensor and then apply the `tf.pad()` function with "VALID" padding.

**Comment by straw-hat-coder** [comment]
Another approach could be to use `tf.tile()` to create multiple copies of the input tensor, concatenate them along the appropriate dimensions, and then use `tf.pad()` with "VALID" padding.

**Comment by Venkat6871** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
